sold.cc

```diff
@@ -378,6 +396,8 @@ void Sold::LoadDynSymtab(ELFBinary* bin, std::vector<Syminfo>& symtab) {
         }
         LOG(INFO) << "Symbol " << name << "@" << bin->name() << " " << sym->st_value;
 
+        p.versym = CalcNewVersym(p.versym, p.sym->st_info);
```

Comment:
Hmm I still don't understand why we can remove non-special version info when constructing the input symbol table (src_syms_). Aren't they necessary to look up the right version?

Reply:
No, I think version info (`Elf_Versym`) is not needed to find the right version
That's because we don't use version info to find a symbol. Instead of it, we use a pair of (soname, version) to find it.

This is the same as glibc. `_dl_lookup_symbol_x` in 
(https://github.com/bminor/glibc/blob/aac0f62c47beee5b546bacc330acc2dd21cda0dc/elf/dl-lookup.c#L830) takes `const struct r_found_version *version` as an argument of version specification. And `r_found_version` defined in (https://github.com/bminor/glibc/blob/aac0f62c47beee5b546bacc330acc2dd21cda0dc/sysdeps/generic/ldsodefs.h#L184) includes no Elf_Versym. `r_found_version` mainly consists of a version name and a filename.

Reply:
Thanks for your explanation! I understood it's safe to update `p.versym` here. However, I'm not fully convinced yet :) Please correct me if I'm wrong, but `Syminfo::versym` is used only by `VersionBuilder::Add` ? If it's the case, cannot we remove this conversion here and

```
void VersionBuilder::Add(Elf_Versym versym, std::string soname, std::string version, StrtabBuilder& strtab) {
    ...
      if (is_special_ver_ndx(versym)) {
        LOG(INFO) << "VersionBuilder::" << special_ver_ndx_to_str(versym);
        vers.push_back(versym);
    } else if {

```

Also, I'm guessing we would not need to keep the original `Elf_Versym` is not necessary to be in Syminfo, but I'm not 100% sure.

Reply:
As for glibc, isn't `_dl_lookup_symbol_x` calling `do_lookup_x`, which calls `check_match`? It looks like this function checks versym?

```
  const ElfW(Half) *verstab = map->l_versyms;
```

and

```
	{
	  /* We can match the version information or use the
	     default one if it is not hidden.  */
	  ElfW(Half) ndx = verstab[symidx] & 0x7fff;
	  if ((map->l_versions[ndx].hash != version->hash
	       || strcmp (map->l_versions[ndx].name, version->name))
	      && (version->hidden || map->l_versions[ndx].hash
		  || (verstab[symidx] & 0x8000)))
	    /* It's not the version we want.  */
	    return NULL;
	}

```

Reply:
It sounds good to move all versym conversion into `VersionBuilder::Add`. I do it in https://github.com/shinh/sold/pull/46/commits/f8f3754efa57bcb7469c4a95f84667da8226a6ab.

Reply:
I don't understand what `verstab[symidx] & 0x8000 != 0`  means. I cannot find any documented information in `elf.h` or something else.

Because it is used in the lookup process, should we use it in `SymtabBuilder::Resolve`?

Reply:
I will do it later.